### PR TITLE
[libc] clean up futex usage

### PR DESCRIPTION
**Pull request by SchrodingerZhu**
# Motivation

Futex syscalls are widely used in our codebase as synchronization mechanism. Hence, it may be worthy to abstract out the most common routines (wait and wake). On the other hand, C++20 also provides `std::atomic_notify_one/std::atomic_wait/std::atomic_notify_all` which align with such functionalities. This PR introduces `Futex` as a subtype of `cpp::Atomic<FutexWordType>` with additional `notify_one/notify_all/wait` operations.

Providing such wrappers also make future porting easier. For example, FreeBSD's `_umtx_op` and Darwin's `ulock` can be wrapped in a similar manner.

### Similar Examples

1. [bionic futex](https://android.googlesource.com/platform/bionic/+/refs/heads/main/libc/bionic/bionic_futex.cpp)
2. [futex in Rust's std](https://github.com/rust-lang/rust/blob/8cef37dbb67e9c80702925f19cf298c4203991e4/library/std/src/sys/pal/unix/futex.rs#L21)

### CODE REVIEW

File: libc/src/__support/threads/linux/futex_utils.h [hunk 1]
```diff
@@ -39,14 +40,26 @@ class Futex : public cpp::Atomic<FutexWordType> {
...
+          /* expected value */ expected,
+          /* timeout */ timeout ? &timeout->abs_time : nullptr,
+          /* ignored */ nullptr,
+          /* bitset */ FUTEX_BITSET_MATCH_ANY);
+
+      // continue waiting if interrupted; otherwise return the result
+      // which should normally be 0 or -ETIMEOUT
+      if (ret == -EINTR)
+        continue;
+      else
```
Comment (lntue) [inline libc/src/__support/threads/linux/futex_utils.h:60]: nit: `else` is not needed.

**Comment by SchrodingerZhu** [comment]
@nickdesaulniers I see both bionic and rust's std library uses only `CLOCK_MONOTONIC` even though `CLOCK_REALTIME` is suggested by POSIX. See below comments from bionic:
```c++
  // pthread's and semaphore's default behavior is to use CLOCK_REALTIME, however this behavior is
  // essentially never intended, as that clock is prone to change discontinuously.
  //
  // What users really intend is to use CLOCK_MONOTONIC, however only pthread_cond_timedwait()
  // provides this as an option and even there, a large amount of existing code does not opt into
  // CLOCK_MONOTONIC.
  //
  // We have seen numerous bugs directly attributable to this difference.  Therefore, we provide
  // this general workaround to always use CLOCK_MONOTONIC for waiting, regardless of what the input
  // timespec is.
  ```

Should we also do something similar? (i.e. not expose realtime clock and document that one should always translate realtime lock to monotonic clock)

**Review by lntue — APPROVED** [review]

**Comment by michaelrj-google** [comment]
I would say that using `CLOCK_MONOTONIC` instead of `CLOCK_REALTIME` is probably a good idea, but should be toggleable with a flag.

**Comment by SchrodingerZhu** [comment]
Then I will leave this as it is. We can leave the responsibility of timespec conversion to outer level.

**Comment by SchrodingerZhu** [comment]
In this case, I will merge this PR as it leaves the API to use realtime lock. Timespec conversion involves `clock_getime` which can be implemented separately in outer-level logic without pulling in `clock_gettime` dependency into `futex_utils`.

**Pull request merged**